presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,304 @@
+    public Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Session session)
+    {
+        if (!(node instanceof JoinNode) || !isJoinReorderingEnabled(session)) {
+            return Optional.empty();
+        }
+
+        JoinNode joinNode = (JoinNode) node;
+        // We check that join distribution type is absent because we only want to do this transformation once (reordered joins will have distribution type already set).
+        if (!(joinNode.getType() == INNER) || joinNode.getDistributionType().isPresent()) {
+            return Optional.empty();
+        }
+
+        MultiJoinNode multiJoinNode = toMultiJoinNode(joinNode, lookup);
+        return Optional.of(new JoinEnumerator(idAllocator, symbolAllocator, session, lookup, multiJoinNode, costComparator).chooseJoinOrder(multiJoinNode));
+    }
+
+    @VisibleForTesting
+    public static class JoinEnumerator
+    {
+        private final Map<MultiJoinNode, JoinNode> memo = new HashMap<>();
+        private final PlanNodeIdAllocator idAllocator;
+        private final Session session;
+        private final Ordering<PlanNode> planNodeOrdering;
+        private final EqualityInference allInference;
+
+        public JoinEnumerator(PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Session session, Lookup lookup, MultiJoinNode multiJoinNode, CostComparator costComparator)
+        {
+            requireNonNull(idAllocator, "idAllocator is null");
+            requireNonNull(symbolAllocator, "symbolAllocator is null");
+            requireNonNull(session, "session is null");
+            requireNonNull(lookup, "lookup is null");
+            requireNonNull(multiJoinNode, "multiJoinNode is null");
+            requireNonNull(costComparator, "costComparator is null");
+            this.idAllocator = idAllocator;
+            this.session = session;
+            this.planNodeOrdering = getPlanNodeOrdering(costComparator, lookup, session, symbolAllocator);
+            this.allInference = createEqualityInference(multiJoinNode.getFilter());
+        }
+
+        private static Ordering<PlanNode> getPlanNodeOrdering(CostComparator costComparator, Lookup lookup, Session session, SymbolAllocator symbolAllocator)
+        {
+            return new Ordering<PlanNode>()
+            {
+                @Override
+                public int compare(PlanNode node1, PlanNode node2)
+                {
+                    PlanNodeCostEstimate node1Cost = lookup.getCumulativeCost(node1, session, symbolAllocator.getTypes());
+                    PlanNodeCostEstimate node2Cost = lookup.getCumulativeCost(node2, session, symbolAllocator.getTypes());
+                    return costComparator.compare(session, node1Cost, node2Cost);
+                }
+            };
+        }
+
+        private JoinNode chooseJoinOrder(MultiJoinNode multiJoinNode)
+        {
+            JoinNode join = memo.get(multiJoinNode);
```

Comment:
What about using the following as memo key:

```
private String makeMemoKey(MultiJoinNode multiJoinNode)
{
    return requireNonNull(multiJoinNode, "multiJoinNode is null")
            .getSources().stream()
            .peek(source -> checkArgument(source instanceof GroupReference, "Source is not an original source: %s", source)) // at least for development time..
            .map(PlanNode::getId) // TODO can original source be non-unique? Can one GroupReference appear in two+ places?
            .map(PlanNodeId::toString)
            .sorted()
            .collect(joining(":")); // assuming ':' cannot appear in node ID
}
```
?

I know this ignores `MultiJoinNode`'s filter, but isn't that a function of sources being joined? I.e. `MultiJoinNode`'s filter = `and( every condition that can be computed based on given sources )`